src/driver.rs

```diff
@@ -410,6 +410,90 @@ impl DriverManager {
         Ok(Driver { c_driver })
     }
 
+    /// Get the Driver based on the file extension from filename
+    ///
+    /// Searches for the available extensions in the registered
+    /// drivers and returns the matches. The determined driver is
+    /// checked for writing capabilities as
+    /// [`Dataset::open`](Dataset::open) can already open datasets
+    /// with just path.
+    ///
+    /// See also: [`get_driver_by_name`](Self::get_driver_by_name)
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::DriverManager;
+    /// # fn main() -> gdal::errors::Result<()> {
+    /// let compatible_drivers =
+    ///     DriverManager::get_drivers_for_filename("test.gpkg", true)
+    ///         .iter()
+    ///         .map(|d| d.short_name())
+    ///         .collect::<Vec<String>>();
+    /// println!("{:?}", compatible_drivers);
+    /// # Ok(())
+    /// # }
+    /// ```
+    /// ```text
+    /// ["GPKG"]
+    /// ```
+    pub fn get_drivers_for_filename(filename: &str, is_vector: bool) -> Vec<Driver> {
+        let ext = {
+            let filename = filename.to_ascii_lowercase();
+            let e = match filename.rsplit_once('.') {
```

Comment:
If we want to ensure true logical parity with GDAL C behavior, we'd use `CPLGetExtension`. 50/50 on whether that's good idea or not.

Reply:
I did consider using the rust's "extension()" method on the `Path` as well, but considering that it can only do the last "." onwards, which requires us to do the `endswith` on the filename for `.shp.zip` and `.gpkg.shp`, I thought just using them directly might be better. And the online protocol addresses with prefix (like `Elasticsearch` in the test) may not be taken as a valid path, so I just did the string thing.

Reply:
It would be nice to have this work for `Path`s. I think you can use `extension` and then `file_stem` and again `extension` to check for this without that much ceremony.

Reply:
`Path` works fine with those, especially as they're just valid file names on Linux. `Path::new` is infallible.

I'd like this because otherwise it's pretty annoying to convert a `Path` or `PathBuf` to a string.

Reply:
That's exactly I don't want to do it lol. 

Like for that: `startswith()` we need to do for "PG:" and similar paths, we'd need to convert the path into string right? And iirc it can't convert it into direct string and has to go through Debug trait and ends up with quoted string or something.

I can try, but unless we can use the `Path` args without having to turn it back to string, it feels like we'll just end up asking for a Path and then internally convert it into string anyway.

Reply:
> And iirc it can't convert it into direct string and has to go through Debug trait and ends up with quoted string or something.

[to_string_lossy](https://doc.rust-lang.org/std/path/struct.Path.html#method.to_string_lossy) will convert it without allocating in the happy case, and without too much ceremony. It's also infallible.

Anyway, I'm not that worried about allocations, just about having a nicer API -- if you have a string, you can get a `Path`, if you have a `Path` it's harder to get a string.

Reply:
> if you have a Path it's harder to get a string.

That conflicts with the first sentence :D

Users can just call `[to_string_lossy](https://doc.rust-lang.org/std/path/struct.Path.html#method.to_string_lossy)`. If it's easy for us, it's easy for them. 

Anyway, I'm trying to do it with `AsRef<Path>`. I'll make a commit if I am successful.

Reply:
> If it's easy for us, it's easy for them.

It's easy for us in the sense that it's infallible, it requires no unsafe code, and it's almost always efficient. It's harder for them because it's ugly and maybe they won't know about `to_string_lossy` so they might try to debug-format them, as you wanted to do.

---

There's also something to be said about being _lossy_, but I guess we only support UTF-8 file names, while GDAL might, in principle, support arbitrary byte sequences). But we hide it, and we could support non-UTF-8 file names today -- at least on Linux -- without changing our API.

On the other hand, if we make the user call `to_string_lossy` on a non-UTF-8 path, they'll get a different name at the output. It's fine when we do it (and also when the user does it) because it won't affect the driver selection in practice. But as you can see, it's a lot of relatively subtle stuff to think about.

Reply:
Haha, I'm currently having that same problem. it seems I can't use Cow<'_, str> on pattern. And might have to do, Variants of `Cow::`. At this point doing if statement seems easier than match. It's really a lot of complications when we have to turn it back into string. :(

Reply:
Okay, let's leave it like this, I'll try to clean it up tomorrow.

Reply:
Because of the abovementioned problem with `Cow`, and not being able to use match properly and so on. I converted it into String after fighting with borrow checking for a while trying to use str.